### PR TITLE
KM478 ✅ Support using AWS profile for authentication

### PR DESCRIPTION
**Pull request by fredriv**
Adds support for authenticating using named AWS profiles and AWS Single Sign-On (SSO).

## Description
This patch enables the user to log in using a named AWS profile, specified in the `AWS_PROFILE` environment variable.

The user can authenticate with IAM access keys without the need for setting secrets in environment variables (instead putting them in the `credentials` file), or authenticate with profiles using AWS SSO or SAML-based login.

Profiles are loaded from the global shared `config` and `credentials` files stored in `~/.aws/` directory, as  opposed to the local config and credentials managed by okctl. 

## Motivation and Context

The change is necessary to support the use of AWS SSO for login.

## How to prove the effect of this PR?

Run `okctl` with the `-a aws-profile` parameter to use AWS profile for authentication. Specify the AWS profile to use in the `AWS_PROFILE` environment variable.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [X] My code follows the code style of this project.
- [X] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [X] All new and existing tests passed.
- [x] I have updated the release notes (for the next release).

### CODE REVIEW

File: pkg/credentials/aws/aws.go [hunk 1]
```diff
@@ -221,6 +224,21 @@ func NewAuthEnvironment(region string, getter KeyGetter) Retriever {
 	}
 }
 
+// NewAuthProfile creates a retriever that fetches credentials from AWS profile
+// environment variable
+func NewAuthProfile(region string, getter KeyGetter) Retriever {
```
Comment (deifyed) [inline pkg/credentials/aws/aws.go:229]: What I meant when I suggested implementing this function was to make it do something like:

```go
// pseudo code
func NewAuthProfile(getter KeyGetter) Retriever {
    profile := getter("AWS_PROFILE")

    credentialsFile := ReadAll("well known credentials file path")

    credentials := make(map[string]Credentials)

    err := ini-library.Unmarshal(credentialsFile, &credentials)

    return &AuthStatic { // or other relevant
        Credentials: credentials[profile]
        IsValid: validate(credentials[profile])
    }
}
```

Doesn't that make sense for this use case?

File: pkg/credentials/aws/aws.go [hunk 2]
```diff
@@ -332,6 +350,7 @@ func (a *AuthSAML) Retrieve() (*Credentials, error) {
 	}
 
 	return &Credentials{
+		AwsProfile:      "default",
```
Comment (deifyed) [inline pkg/credentials/aws/aws.go:353]: This should probably be a constant somewhere. `pkg/config/constant/api.go` is a good place if its needed globally

**Review by deifyed — APPROVED** [review]
I still feel like it should be able to funnel all three different ways of authenticating into one streamlined way of relating to authentication in okctl centralising around profiles, but if it works I guess we could clean it up later

Anyways, good job! Authentication-meddling as the second real PR is what I consider an ilddåp. Looks good

File: pkg/credentials/aws/aws.go [hunk 3]
```diff
@@ -616,6 +635,7 @@ func (s *IniPersister) Get() (*Credentials, error) {
 	}
 
 	return &Credentials{
+		AwsProfile:      "default", // SAML login uses 'default' profile
```
[the remark this answers is not in the document]
Comment (fredriv) [reply pkg/credentials/aws/aws.go:638]: :+1:

**Comment by fredriv** [comment]
@deifyed Pushed so changes based on your comments. Thanks! :-)

File: pkg/commands/venv_test.go [hunk 4]
```diff
@@ -47,6 +50,53 @@ func TestCleanOsEnvVars(t *testing.T) {
...
+
+//nolint:errcheck
+func TestGetOkctlEnvVarsForAwsProfile(t *testing.T) {
+	okctlEnv := OkctlEnvironment{
+		AWSCredentialsType: context.AWSCredentialsTypeAwsProfile,
+	}
+	os.Setenv("AWS_PROFILE", "testprofile")
+	os.Setenv("HOME", "/home/user")
+
+	envVars, err := GetVenvEnvVars(okctlEnv)
```
[the remark this answers is not in the document]
Comment (fredriv) [reply pkg/commands/venv_test.go]: Turns out we're calling the `os.UserHomeDir()` function which reads the `$HOME` environment variable, so this didn't work ;-) Instead implemented a helper function `SetOsEnv` in the `testing` package which helps with setting and resetting the environment variables.

[the remark this answers is not in the document]
Comment (yngvark) [reply pkg/commands/venv_test.go]: I recommend to pass these values through `opts OkctlEnvironment` (unless it somehow isn't possible). The intention of this struct is to pass all needed data. So the `os` function calls will then be placed in `commands.GetOkctlEnvironment`. 

As a general rule, I also recommend not setting OS environment variables in tests, unless necessary. It has the same issues as settings global variables, hindring tests to be run in parallell, for instance.

**Review by yngvark — CHANGES_REQUESTED** [review]
Really good research & development! Auth isn't easy.

See my comment on the thread regarding GetVenvEnvVars, I have a suggestion for where to put the `os` function calls.

**Review by yngvark — APPROVED** [review]
Looks good :+1

**Pull request merged**